### ISSUE TITLE
Is it possible to write Analysers and Refactoring like it is done for Roslyn in VS2015

### ISSUE THREAD

**Issue by xperiandri** [Microsoft/visualfsharp#541]
Is it possible to write Analyzers and Refactorings using F# compiler data like it is done for C# and VB in Visual Studio 2015 and integrate them with Light Bulb feature of Visual Studio 2015?

**Comment by latkin** [comment]
No, the standard Visual Studio F# tools/compiler stack does not expose code analysis APIs like Roslyn does.

The [F# Compiler Services](http://fsharp.github.io/FSharp.Compiler.Service/) project (not part of the standard VS install, but well-supported by the community) does offer an API layer similar to Roslyn.  FCS is used by the [Visual F# Power Tools](https://github.com/fsprojects/VisualFSharpPowerTools) extension (another well-supported community project), which does indeed offer more advanced refactorings and code analysis.  Although VFPT isn't "extensible," it is open source so any analysis/refactorings you have in mind could be integrated into their codebase directly.

**Comment by xperiandri** [comment]
How to get started?
F# beginners need a guide!
Can you create a Project Templates similar to https://visualstudiogallery.msdn.microsoft.com/2ddb7240-5249-4c8c-969e-5d05823bcb89?SRC=VSIDE
So everyone can quickly get started

**Comment by xperiandri** [comment]
@dsyme what do you think about my question?

**Comment by dsyme** [comment]
@xperiandri  I think the best corresponding thing in the F# world is to contribute to the [FSharpLint](https://github.com/fsprojects/FSharpLint) or [Visual F# Power Tools](https://github.com/fsprojects/VisualFSharpPowerTools) projects.  

cheers!
don

**Comment by xperiandri** [comment]
It is too complex. I have to understand how that work, than do a pull request.
And what if I want my own set of analyzers?
I feel with F# like back to 90's: no syntax highlighting, no navigation bar, no refactorings, no snippets, no code lens, no analyzers - nothing! Notepad is your best friend...
How can I tell people to use F# is it is a whole pain?

**Comment by dsyme** [comment]
@xperiandri,  [The Visual F# Power Tools](http://fsprojects.github.io/VisualFSharpPowerTools/) have the features listed below. Have you installed those?

* Auto-generating XmlDoc
* Source code formatting
* Navigation bar
* Highlight references
* Rename refactoring
* Depth colorizer
* NavigateTo
* Syntax coloring
* Folder organization
* Find all references
* Implement interface
* Record stub generation
* Union pattern match case generation
* Resolve unopened namespaces and modules
* Go to metadata
* Task List comments
* Generate references for F# Interactive
* Navigate to source
* Quick info panel

**Comment by xperiandri** [comment]
Yes I did
Without Power Tools coding would be impossible

**Comment by forki** [comment]
Strange. I must do something wrong. I mean I really really like VFPT, but even before that tool existed I felt much more productive in F# than in C# + R#. Maybe I'm just a 90's guy...

**Comment by vasily-kirichenko** [comment]
What's more, we started building VFPT in bare VS! Guess how impossibly hard it was :)

**Comment by xperiandri** [comment]
But we are in 2015 now when C# and VB coders have advantage in the form of Visual Studio project templates for analyzers.

Is it too hard to do the same for F# coders?
Now instead of implementing a piece of code to do some refactoring I have to learn how to integrate it into VFPT and need to learn lots of Visual Studio internals...

**Comment by vasily-kirichenko** [comment]
1. Yes, it's hard.
2. MS F# team is: two (partially working on F#) developers + a PM.
3. I'd not expect significant improving either tooling nor the language in near future (1-2 years).
4. I you don't like how things state, you can: pick another language, like Scala or C#, or contribute to F#.

**Comment by forki** [comment]
What analyzers do you want to write? AFAIK VFPT gets the AST from the FSharp.Compiler.Service. So I think it should still be possible. Probably you will have to create a bit of additional infrastructure  in the project, but it would be a great addition.

Sure roslyn is very streamlined in this regard, but lots of MS people worked on it for a very very long time. In F# we have to do similar things in the community, but we can still achieve great things.

**Comment by xperiandri** [comment]
I will have a look at how exactly it is done in Roslyn.
You understood me absolutely right, I want an infrastructure for F# analyzers to be build.
I suppose that it is possible to reuse major part of Roslyn for this.

**Comment by xperiandri** [comment]
Is there any good manual about F# compiler?

**Comment by forki** [comment]
There is currently no wrapper from FCS AST to roslyn Ast. IIRC there were investigations in that area? /cc @7sharp9

**Comment by 7sharp9** [comment]
F# has several AST representations see [here](http://7sharpnine.com/posts/meta-matic/)

You could map the untyped or the typed AST to a roslyn AST, but it would be significant work.

**Comment by forki** [comment]
@xperiandri http://fsharp.org/specs/language-spec/ ?

**Comment by xperiandri** [comment]
Merged discussion from https://github.com/Microsoft/visualfsharp/issues/540#issuecomment-124259274

@vladima, so I read that links and also [this one](https://msdn.microsoft.com/en-us/library/vstudio/dn903708(v=vs.140).aspx) from @dungpa and a few more about this and about Roslyn and its analyzers.

So I came to this question https://github.com/dotnet/roslyn/issues/4520
Because what I want to achieve does not look to hard.
Actually VFPT currently does something similar to Roslyn analyzers but without common infrastructure.

Don't you think that building such infrastructure would be more effective than doing analysis and displaying UI like it is done currently?

I mean that analyzer is just a definition of filer and provider of actions and changes.
I subscribe for particular syntax construct and when Roslyn analyzer is invoked it just provides a set of actions possible at that construct and presents a look of changes.

So @vasily-kirichenko [here](https://github.com/fsprojects/VisualFSharpPowerTools/issues/1049) does similar things. But he filters syntax constructs himself. And does not use API that generates a preview.
So to "mimic Roslyn analyzer" = "create any refactoring for F#" we need to do the same stuff of filtering and UI creating.
